File: fused/computer.go
package main

// The computer control surface: the guest half of Anthropic's computer use
// tool. Claude never reaches this directly — the caller's agent loop receives
// a tool_use block, turns it into one POST here, and returns the screenshot
// as the tool_result. fused only executes actions against the local display.
//
// Everything in an action is untrusted input that becomes a process argv.
// Invocations are built as argv slices and executed directly — no shell, so
// no element can be reinterpreted as a pipeline or a redirect — and key names
// are held to a charset, the same treatment the env-key validation in
// internal/fusefile/compile.go gives identifiers that reach a file the guest
// sources.
//
// The routes are registered on every image. On a non-desktop image the
// display socket is absent and every call answers 503 with a reason, which is
// what makes them safe to serve unconditionally.

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"image"
	"image/draw"
	"image/png"
	"net/http"
	"os"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
	"time"
)

const (
	// maxComputerBody bounds the action request; the largest legitimate field
	// is a typed string, and 1MB is far beyond any of those.
	maxComputerBody = 1 << 20
	// maxTextLen bounds type/key payloads.
	maxTextLen = 32768
	// maxCoordinate bounds any coordinate or region component; no supported
	// display is anywhere near this large.
	maxCoordinate = 10000
	// maxHoldSeconds bounds hold_key and wait durations.
	maxHoldSeconds = 100
	// settleDelay is how long an input action gets to repaint before the
	// screenshot that rides back on its response is captured.
	settleDelay = 300 * time.Millisecond
	// execTimeout bounds a single xdotool/scrot invocation.
	execTimeout = 10 * time.Second
)

// keyPattern admits xdotool keysym combos ("ctrl+shift+t", "Return",
// "XF86AudioMute"). Anything else — spaces, quotes, path separators — is
// rejected before it can become an argv element.
var keyPattern = regexp.MustCompile(`^[A-Za-z0-9_+]{1,128}$`)

// coordinate is an x/y pixel pair. on the wire it stays the two-element
// [x, y] array Claude's tool_use input carries; only the go side gets named
// fields.
type coordinate struct {
	X int
	Y int
}

func (c coordinate) MarshalJSON() ([]byte, error) {
	return json.Marshal([2]int{c.X, c.Y})
}

func (c *coordinate) UnmarshalJSON(data []byte) error {
	var pair []int
	if err := json.Unmarshal(data, &pair); err != nil {
		return err
	}
	if len(pair) != 2 {
		return fmt.Errorf("want [x, y], got %d elements", len(pair))
	}
	c.X, c.Y = pair[0], pair[1]
	return nil
}

// computerAction is the request body of POST /v1/computer/action. The field
// names mirror the tool_use input Claude emits so a caller's translation
// layer stays mechanical.
type computerAction struct {
	Action          string      `json:"action"`
	Coordinate      *coordinate `json:"coordinate,omitempty"`
	StartCoordinate *coordinate `json:"start_coordinate,omitempty"`
	Text            string      `json:"text,omitempty"`
	Region          []int       `json:"region,omitempty"`
	Duration        float64     `json:"duration,omitempty"`
	ScrollDirection string      `json:"scroll_direction,omitempty"`
	ScrollAmount    int         `json:"scroll_amount,omitempty"`
}

// computerResult is the response body. Screenshot is a base64 PNG, present on
// every action that implies one.
type computerResult struct {
	Output     string `json:"output,omitempty"`
	Screenshot string `json:"screenshot,omitempty"`
}

// scrollButtons maps a scroll direction to the X button that emits it.
var scrollButtons = map[string]string{
	"up":    "4",
	"down":  "5",
	"left":  "6",
	"right": "7",
}

// computer executes actions against one X display. run and ready are fields
// so tests can assert the argv an action compiles to, and drive the handlers,
// without an X server present.
type computer struct {
	display string
	// vncAddr is where the stream route finds the local vnc server; a field
	// so tests can point it at a stub listener.
	vncAddr string
	run     func(ctx context.Context, argv []string) (string, error)
	ready   func() error
}

func newComputer(display string) *computer {
	if display == "" {
		display = ":1"
	}
	c := &computer{display: display, vncAddr: vncAddr}
	c.run = c.execRun
	c.ready = c.checkReady
	return c
}

// socketPath returns the unix socket the display serves on, derived from the
// display string (":1" or ":1.0" -> /tmp/.X11-unix/X1).
func (c *computer) socketPath() string {
	num := strings.TrimPrefix(c.display, ":")
	if i := strings.IndexByte(num, '.'); i >= 0 {
		num = num[:i]
	}
	return "/tmp/.X11-unix/X" + num
}

// checkReady reports why the surface cannot serve, or nil when it can. The
// checks name the missing piece so a caller on a non-desktop image learns
// what is absent rather than getting a bare 503.
func (c *computer) checkReady() error {
	if _, err := exec.LookPath("xdotool"); err != nil {
		return fmt.Errorf("xdotool not installed; this is not a desktop image")
	}
	if _, err := exec.LookPath("scrot"); err != nil {
		return fmt.Errorf("scrot not installed; this is not a desktop image")
	}
	if _, err := os.Stat(c.socketPath()); err != nil {
		return fmt.Errorf("display %s is not up (no socket at %s)", c.display, c.socketPath())
	}
	return nil
}

// execRun executes one argv against the display and returns its stdout.
func (c *computer) execRun(ctx context.Context, argv []string) (string, error) {
	ctx, cancel := context.WithTimeout(ctx, execTimeout)
	defer cancel()
	cmd := exec.CommandContext(ctx, argv[0], argv[1:]...)
	cmd.Env = append(os.Environ(), "DISPLAY="+c.display)
	var out, errb bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = &errb
	if err := cmd.Run(); err != nil {
		msg := strings.TrimSpace(errb.String())
		if msg == "" {
			msg = err.Error()
		}
		return "", fmt.Errorf("%s: %s", argv[0], msg)
	}
	return out.String(), nil
}

// capturePNG takes a screenshot of the whole display and returns the PNG
// bytes. scrot writes to a file, so the capture goes through a temp path that
// is removed either way.
func (c *computer) capturePNG(ctx context.Context) ([]byte, error) {
	f, err := os.CreateTemp("", "fused-screenshot-*.png")
	if err != nil {
		return nil, err
	}
	path := f.Name()
	_ = f.Close()
	defer func() { _ = os.Remove(path) }()

	if _, err := c.run(ctx, []string{"scrot", "-o", path}); err != nil {
		return nil, err
	}
	b, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	if len(b) == 0 {
		return nil, fmt.Errorf("screenshot capture produced an empty file")
	}
	return b, nil
}

// geometry asks the display for its size.
func (c *computer) geometry(ctx context.Context) (w, h int, err error) {
	out, err := c.run(ctx, []string{"xdotool", "getdisplaygeometry"})
	if err != nil {
		return 0, 0, err
	}
	fields := strings.Fields(out)
	if len(fields) != 2 {
		return 0, 0, fmt.Errorf("unexpected getdisplaygeometry output %q", out)
	}
	w, err = strconv.Atoi(fields[0])
	if err == nil {
		h, err = strconv.Atoi(fields[1])
	}
	return w, h, err
}

// actionError is a request-shape failure that should answer 400 rather
// than 500.
type actionError struct{ msg string }

func (e *actionError) Error() string { return e.msg }

func badAction(format string, args ...any) error {
	return &actionError{msg: fmt.Sprintf(format, args...)}
}

// validCoordinate checks an [x, y] pair.
func validCoordinate(name string, c *coordinate) error {
	if c == nil {
		return badAction("%s: want [x, y], got nothing", name)
	}
	for _, v := range []int{c.X, c.Y} {
		if v < 0 || v > maxCoordinate {
			return badAction("%s: %d out of range 0..%d", name, v, maxCoordinate)
		}
	}
	return nil
}

// validKeys checks a keysym combo the way env keys are checked before they
// reach a file the guest sources: charset first, questions later.
func validKeys(text string) error {
	if !keyPattern.MatchString(text) {
		return badAction("key: %q is not a valid keysym combo (letters, digits, _ and + only)", text)
	}
	return nil
}

// moveArgs returns the chained xdotool words that position the pointer, or
// nothing when the action did not name a coordinate.
func moveArgs(c *coordinate) []string {
	if c == nil {
		return nil
	}
	return []string{"mousemove", "--sync", strconv.Itoa(c.X), strconv.Itoa(c.Y)}
}

// clickButtons maps the click actions to their X button and repeat count.
var clickActions = map[string]struct {
	button string
	repeat int
}{
	"left_click":   {"1", 1},
	"right_click":  {"3", 1},
	"middle_click": {"2", 1},
	"double_click": {"1", 2},
	"triple_click": {"1", 3},
}

// perform executes one decoded action and reports its textual output (if
// any) and whether a screenshot should ride back on the response.
func (c *computer) perform(ctx context.Context, a computerAction) (output string, screenshot bool, err error) {
	switch a.Action {
	case "screenshot":
		return "", true, nil

	case "cursor_position":
		out, err := c.run(ctx, []string{"xdotool", "getmouselocation"})
		if err != nil {
			return "", false, err
		}
		return strings.TrimSpace(out), false, nil

	case "mouse_move":
		if err := validCoordinate("coordinate", a.Coordinate); err != nil {
			return "", false, err
		}
		_, err := c.run(ctx, append([]string{"xdotool"}, moveArgs(a.Coordinate)...))
		return "", true, err

	case "left_click", "right_click", "middle_click", "double_click", "triple_click":
		spec := clickActions[a.Action]
		if a.Coordinate != nil {
			if err := validCoordinate("coordinate", a.Coordinate); err != nil {
				return "", false, err
			}
		}
		// optional held modifiers ride in text ("ctrl+shift"); they are
		// pressed, the click happens, and the release is guaranteed even
		// when the click itself fails, so a failed action cannot leave the
		// session with a stuck modifier.
		if a.Text != "" {
			if err := validKeys(a.Text); err != nil {
				return "", false, err
			}
			if _, err := c.run(ctx, []string{"xdotool", "keydown", "--", a.Text}); err != nil {
				return "", false, err
			}
			defer func() { _, _ = c.run(ctx, []string{"xdotool", "keyup", "--", a.Text}) }()
		}
		argv := append([]string{"xdotool"}, moveArgs(a.Coordinate)...)
		argv = append(argv, "click")
		if spec.repeat > 1 {
			argv = append(argv, "--repeat", strconv.Itoa(spec.repeat), "--delay", "10")
		}
		argv = append(argv, spec.button)
		_, err := c.run(ctx, argv)
		return "", true, err

	case "left_click_drag":
		if err := validCoordinate("start_coordinate", a.StartCoordinate); err != nil {
			return "", false, err
		}
		if err := validCoordinate("coordinate", a.Coordinate); err != nil {
			return "", false, err
		}
		argv := append([]string{"xdotool"}, moveArgs(a.StartCoordinate)...)
		argv = append(argv, "mousedown", "1")
		argv = append(argv, moveArgs(a.Coordinate)...)
		argv = append(argv, "mouseup", "1")
		_, err := c.run(ctx, argv)
		return "", true, err

	case "left_mouse_down", "left_mouse_up":
		verb := "mousedown"
		if a.Action == "left_mouse_up" {
			verb = "mouseup"
		}
		if a.Coordinate != nil {
			if err := validCoordinate("coordinate", a.Coordinate); err != nil {
				return "", false, err
			}
		}
		argv := append([]string{"xdotool"}, moveArgs(a.Coordinate)...)
		argv = append(argv, verb, "1")
		_, err := c.run(ctx, argv)
		return "", true, err

	case "type":
		if a.Text == "" {
			return "", false, badAction("type: text is required")
		}
		if len(a.Text) > maxTextLen {
			return "", false, badAction("type: text exceeds %d bytes", maxTextLen)
		}
		// the payload sits behind `--` as a single argv element, so it is
		// typed verbatim whatever it contains.
		_, err := c.run(ctx, []string{"xdotool", "type", "--delay", "12", "--", a.Text})
		return "", true, err

	case "key":
		if a.Text == "" {
			return "", false, badAction("key: text is required")
		}
		if err := validKeys(a.Text); err != nil {
			return "", false, err
		}
		_, err := c.run(ctx, []string{"xdotool", "key", "--", a.Text})
		return "", true, err

	case "hold_key":
		if a.Text == "" {
			return "", false, badAction("hold_key: text is required")
		}
		if err := validKeys(a.Text); err != nil {
			return "", false, err
		}
		if a.Duration <= 0 || a.Duration > maxHoldSeconds {
			return "", false, badAction("hold_key: duration must be in (0, %d] seconds", maxHoldSeconds)
		}
		if _, err := c.run(ctx, []string{"xdotool", "keydown", "--", a.Text}); err != nil {
			return "", false, err
		}
		select {
		case <-time.After(time.Duration(a.Duration * float64(time.Second))):
		case <-ctx.Done():
		}
		_, err := c.run(ctx, []string{"xdotool", "keyup", "--", a.Text})
		return "", true, err

	case "scroll":
		button, ok := scrollButtons[a.ScrollDirection]
		if !ok {
			return "", false, badAction("scroll: scroll_direction must be up, down, left or right")
		}
		if a.ScrollAmount <= 0 || a.ScrollAmount > 100 {
			return "", false, badAction("scroll: scroll_amount must be in 1..100")
		}
		if a.Coordinate != nil {
			if err := validCoordinate("coordinate", a.Coordinate); err != nil {
				return "", false, err
			}
		}
		argv := append([]string{"xdotool"}, moveArgs(a.Coordinate)...)
		argv = append(argv, "click", "--repeat", strconv.Itoa(a.ScrollAmount), button)
		_, err := c.run(ctx, argv)
		return "", true, err

	case "wait":
		if a.Duration <= 0 || a.Duration > maxHoldSeconds {
			return "", false, badAction("wait: duration must be in (0, %d] seconds", maxHoldSeconds)
		}
		select {
		case <-time.After(time.Duration(a.Duration * float64(time.Second))):
		case <-ctx.Done():
		}
		return "", true, nil

	case "zoom":
		// handled by the caller: it needs the cropped screenshot itself
		// rather than a full capture appended afterwards.
		return "", false, badAction("zoom: internal dispatch error")

	case "":
		return "", false, badAction("action is required")

	default:
		return "", false, badAction("unknown action %q", a.Action)
	}
}

// cropPNG decodes a full-display capture and returns the region
// [x1, y1, x2, y2] re-encoded at full resolution, which is exactly what the
// zoom action of computer_20251124 expects.
func cropPNG(full []byte, region []int) ([]byte, error) {
	if len(region) != 4 {
		return nil, badAction("zoom: region wants [x1, y1, x2, y2], got %d elements", len(region))
	}
	for _, v := range region {
		if v < 0 || v > maxCoordinate {
			return nil, badAction("zoom: region value %d out of range 0..%d", v, maxCoordinate)
		}
	}
	x1, y1, x2, y2 := region[0], region[1], region[2], region[3]
	if x2 <= x1 || y2 <= y1 {
		return nil, badAction("zoom: region must have positive width and height")
	}
	img, err := png.Decode(bytes.NewReader(full))
	if err != nil {
		return nil, fmt.Errorf("decode screenshot: %w", err)
	}
	rect := image.Rect(x1, y1, x2, y2).Intersect(img.Bounds())
	if rect.Empty() {
		return nil, badAction("zoom: region lies outside the display")
	}
	out := image.NewRGBA(image.Rect(0, 0, rect.Dx(), rect.Dy()))
	draw.Draw(out, out.Bounds(), img, rect.Min, draw.Src)
	var buf bytes.Buffer
	if err := png.Encode(&buf, out); err != nil {
		return nil, fmt.Errorf("encode zoom region: %w", err)
	}
	return buf.Bytes(), nil
}

// handleAction serves POST /v1/computer/action.
func (c *computer) handleAction(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}
	if err := c.ready(); err != nil {
		writeJSON(w, http.StatusServiceUnavailable, map[string]string{"error": err.Error()})
		return
	}
	var a computerAction
	body := http.MaxBytesReader(w, r.Body, maxComputerBody)
	if err := json.NewDecoder(body).Decode(&a); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "malformed action: " + err.Error()})
		return
	}

	// zoom is its own path: the screenshot is the action.
	if a.Action == "zoom" {
		full, err := c.capturePNG(r.Context())
		if err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
			return
		}
		cropped, err := cropPNG(full, a.Region)
		if err != nil {
			writeActionError(w, err)
			return
		}
		writeJSON(w, http.StatusOK, computerResult{Screenshot: base64.StdEncoding.EncodeToString(cropped)})
		return
	}

	output, wantShot, err := c.perform(r.Context(), a)
	if err != nil {
		writeActionError(w, err)
		return
	}
	res := computerResult{Output: output}
	if wantShot {
		if a.Action != "screenshot" {
			time.Sleep(settleDelay)
		}
		shot, err := c.capturePNG(r.Context())
		if err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
			return
		}
		res.Screenshot = base64.StdEncoding.EncodeToString(shot)
	}
	writeJSON(w, http.StatusOK, res)
}

// handleScreenshot serves GET /v1/computer/screenshot, the cheap read-only
// path for callers that want pixels without side effects.
func (c *computer) handleScreenshot(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}
	if err := c.ready(); err != nil {
		writeJSON(w, http.StatusServiceUnavailable, map[string]string{"error": err.Error()})
		return
	}
	shot, err := c.capturePNG(r.Context())
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
		return
	}
	writeJSON(w, http.StatusOK, computerResult{Screenshot: base64.StdEncoding.EncodeToString(shot)})
}

// handleDisplay serves GET /v1/computer/display: geometry and liveness, so a
// caller can populate display_width_px / display_height_px in its tool
// definition without hardcoding them. Unlike the action routes this answers
// 200 with up=false when there is no display, because "is there a desktop" is
// exactly the question it exists to answer.
func (c *computer) handleDisplay(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}
	body := map[string]any{"display": c.display, "up": false, "width": 0, "height": 0}
	if err := c.ready(); err != nil {
		body["error"] = err.Error()
		writeJSON(w, http.StatusOK, body)
		return
	}
	wpx, hpx, err := c.geometry(r.Context())
	if err != nil {
		body["error"] = err.Error()
		writeJSON(w, http.StatusOK, body)
		return
	}
	body["up"] = true
	body["width"] = wpx
	body["height"] = hpx
	writeJSON(w, http.StatusOK, body)
}

// writeActionError maps request-shape failures to 400 and everything else
// (xdotool failures, capture failures) to 500.
func writeActionError(w http.ResponseWriter, err error) {
	var ae *actionError
	if errors.As(err, &ae) {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": err.Error()})
		return
	}
	writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
}

File: sdks/rust/src/computer.rs
//! The computer-use adapter: the glue between a Claude agent loop and an
//! environment's desktop.
//!
//! Computer use is a client-side tool. Claude never connects to the sandbox:
//! it emits a `tool_use` block, the caller's loop executes it against an
//! environment it controls, and returns a `tool_result` with a screenshot.
//! This module is that translation, done once, so nobody has to rediscover
//! that the `tool_use` input is already the wire shape the computer endpoint
//! takes.

use reqwest::Method;
use serde::{Deserialize, Serialize};

use crate::environments::Environments;
use crate::error::Error;
use crate::strenum::string_enum;
use crate::transport::require;

string_enum! {
    /// Computer-use actions the guest agent understands, in the same
    /// vocabulary Anthropic's computer tool emits.
    pub enum ComputerActionKind {
        Screenshot => "screenshot",
        CursorPosition => "cursor_position",
        MouseMove => "mouse_move",
        LeftClick => "left_click",
        RightClick => "right_click",
        MiddleClick => "middle_click",
        DoubleClick => "double_click",
        TripleClick => "triple_click",
        LeftClickDrag => "left_click_drag",
        LeftMouseDown => "left_mouse_down",
        LeftMouseUp => "left_mouse_up",
        /// Types a string of text (the tool's `type` action).
        Type => "type",
        Key => "key",
        HoldKey => "hold_key",
        Scroll => "scroll",
        Wait => "wait",
        Zoom => "zoom",
    }
}

string_enum! {
    /// Directions for [`ComputerActionKind::Scroll`].
    pub enum ScrollDirection {
        Up => "up",
        Down => "down",
        Left => "left",
        Right => "right",
    }
}

/// An `[x, y]` pixel position on the display.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct Coordinate(pub u32, pub u32);

/// An `[x, y, width, height]` crop of the display, for
/// [`ComputerActionKind::Zoom`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct Region(pub u32, pub u32, pub u32, pub u32);

/// One computer-use action, in the same shape Anthropic's computer tool
/// emits as `tool_use` input, so translating a `tool_use` block into a call
/// is mechanical. Only `action` is required; which other fields apply
/// depends on the action, and the guest agent enforces the full schema.
///
/// ```
/// use fuse::{ComputerAction, ScrollDirection};
///
/// let shot = ComputerAction::screenshot();
/// let click = ComputerAction::left_click(640, 400);
/// let scroll = ComputerAction::scroll(640, 400, ScrollDirection::Down, 3);
/// ```
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ComputerAction {
    pub action: ComputerActionKind,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub coordinate: Option<Coordinate>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub start_coordinate: Option<Coordinate>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub region: Option<Region>,
    /// Seconds, for [`ComputerActionKind::Wait`] and
    /// [`ComputerActionKind::HoldKey`].
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub duration: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub scroll_direction: Option<ScrollDirection>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub scroll_amount: Option<u32>,
}

impl ComputerAction {
    /// Returns a bare action of the given kind; chain setters for its
    /// parameters.
    pub fn new(action: ComputerActionKind) -> Self {
        Self {
            action,
            coordinate: None,
            start_coordinate: None,
            text: None,
            region: None,
            duration: None,
            scroll_direction: None,
            scroll_amount: None,
        }
    }

    /// Captures the display.
    pub fn screenshot() -> Self {
        Self::new(ComputerActionKind::Screenshot)
    }

    /// Left-clicks at `(x, y)`.
    pub fn left_click(x: u32, y: u32) -> Self {
        Self::new(ComputerActionKind::LeftClick).coordinate(x, y)
    }

    /// Types a string of text.
    pub fn type_text(text: impl Into<String>) -> Self {
        Self::new(ComputerActionKind::Type).text(text)
    }

    /// Presses a key or key combination (e.g. `Return`, `ctrl+s`).
    pub fn key(text: impl Into<String>) -> Self {
        Self::new(ComputerActionKind::Key).text(text)
    }

    /// Scrolls at `(x, y)` by `amount` clicks in `direction`.
    pub fn scroll(x: u32, y: u32, direction: ScrollDirection, amount: u32) -> Self {
        let mut action = Self::new(ComputerActionKind::Scroll).coordinate(x, y);
        action.scroll_direction = Some(direction);
        action.scroll_amount = Some(amount);
        action
    }

    pub fn coordinate(mut self, x: u32, y: u32) -> Self {
        self.coordinate = Some(Coordinate(x, y));
        self
    }

    pub fn start_coordinate(mut self, x: u32, y: u32) -> Self {
        self.start_coordinate = Some(Coordinate(x, y));
        self
    }

    pub fn text(mut self, text: impl Into<String>) -> Self {
        self.text = Some(text.into());
        self
    }

    pub fn region(mut self, x: u32, y: u32, width: u32, height: u32) -> Self {
        self.region = Some(Region(x, y, width, height));
        self
    }

    pub fn duration(mut self, seconds: f64) -> Self {
        self.duration = Some(seconds);
        self
    }
}

/// The guest's answer to one action. `screenshot` is a base64 PNG, present
/// on every action that implies one; for zoom it is the cropped region at
/// full resolution.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct ComputerResult {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub screenshot: Option<String>,
}

/// Reports the environment's display: whether it is up and at what
/// geometry, so a caller can populate `display_width_px` /
/// `display_height_px` in its computer tool definition without hardcoding
/// them.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct ComputerDisplay {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub display: Option<String>,
    pub up: bool,
    pub width: u32,
    pub height: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// A Messages-API base64 image source.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ImageSource {
    /// Always `base64`.
    #[serde(rename = "type")]
    pub kind: String,
    /// Always `image/png`.
    pub media_type: String,
    pub data: String,
}

impl ImageSource {
    /// Returns a base64 PNG source.
    pub fn png(data: impl Into<String>) -> Self {
        Self {
            kind: "base64".to_owned(),
            media_type: "image/png".to_owned(),
            data: data.into(),
        }
    }
}

/// One content block of a Messages API `tool_result`. The serialized shape
/// matches the Messages API, so a caller can serialize the content straight
/// into the `tool_result` they send back.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ToolResultBlock {
    Text { text: String },
    Image { source: ImageSource },
}

/// What goes back to Claude for one computer `tool_use`: the content blocks
/// and whether they describe an error. Map it onto the `tool_result` block
/// for the `tool_use`'s id and the loop is closed.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ComputerToolResult {
    pub content: Vec<ToolResultBlock>,
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub is_error: bool,
}

fn tool_error(message: &str) -> ComputerToolResult {
    let message = if message.is_empty() {
        "the computer action failed"
    } else {
        message
    };
    ComputerToolResult {
        is_error: true,
        content: vec![ToolResultBlock::Text {
            text: message.to_owned(),
        }],
    }
}

impl Environments {
    /// Relays one computer-use action to the environment's desktop and
    /// returns the result, usually carrying a screenshot. It requires an
    /// environment booted from a desktop image; on any other image the
    /// server answers 503 with a reason. Screenshots ride back
    /// base64-encoded, so the no-timeout stream client carries the call.
    pub async fn computer(
        &self,
        vm_id: &str,
        action: ComputerAction,
    ) -> Result<ComputerResult, Error> {
        require(vm_id, "vm id")?;
        let request = self
            .transport()
            .stream_request(Method::POST, &["v1", "environments", vm_id, "computer"])
            .json(&action);
        self.transport().send_json(request).await
    }

    /// Reports whether the environment has a live display and at what
    /// geometry. `up` is false with a reason on an image with no desktop.
    pub async fn computer_display(&self, vm_id: &str) -> Result<ComputerDisplay, Error> {
        require(vm_id, "vm id")?;
        let request = self
            .transport()
            .request(Method::GET, &["v1", "environments", vm_id, "computer"]);
        self.transport().send_json(request).await
    }

    /// Executes one computer `tool_use` against an environment and shapes
    /// the answer for the Messages API.
    ///
    /// `input` is the `tool_use` block's input field, verbatim: it is
    /// already the shape the computer endpoint takes, and it is forwarded as
    /// raw JSON rather than re-encoded, so action fields this SDK has not
    /// learned yet still reach the guest.
    ///
    /// An action the environment rejects (a malformed coordinate, a display
    /// that is not up) comes back as `is_error` content rather than an
    /// `Err`: the agent loop should feed it to the model, which is the party
    /// that can correct it. An `Err` is reserved for the failures the model
    /// cannot fix — transport, auth, an unknown environment.
    pub async fn computer_tool_result(
        &self,
        vm_id: &str,
        input: &serde_json::Value,
    ) -> Result<ComputerToolResult, Error> {
        require(vm_id, "vm id")?;
        let action = match input {
            serde_json::Value::Object(fields) => fields
                .get("action")
                .and_then(serde_json::Value::as_str)
                .unwrap_or_default(),
            _ => {
                return Err(Error::InvalidRequest(
                    "tool input is not a json object".to_owned(),
                ));
            }
        };
        if action.is_empty() {
            return Ok(tool_error("tool input names no action"));
        }

        let request = self
            .transport()
            .stream_request(Method::POST, &["v1", "environments", vm_id, "computer"])
            .json(input);
        let result: ComputerResult = match self.transport().send_json(request).await {
            Ok(result) => result,
            Err(Error::Api(api_err)) if matches!(api_err.status, 400 | 409 | 503) => {
                // the model asked for something the environment refused;
                // telling the model is the fix, not failing the loop.
                return Ok(tool_error(&api_err.message));
            }
            Err(err) => return Err(err),
        };

        let mut content = Vec::new();
        if let Some(output) = result.output.filter(|text| !text.is_empty()) {
            content.push(ToolResultBlock::Text { text: output });
        }
        if let Some(screenshot) = result.screenshot.filter(|data| !data.is_empty()) {
            content.push(ToolResultBlock::Image {
                source: ImageSource::png(screenshot),
            });
        }
        if content.is_empty() {
            // an action with nothing to show still needs a tool_result.
            content.push(ToolResultBlock::Text {
                text: "done".to_owned(),
            });
        }
        Ok(ComputerToolResult {
            content,
            is_error: false,
        })
    }
}
